### PR TITLE
[SPARK-26742][K8S] Update Kubernetes-Client version to 4.1.2

### PR DESCRIPTION
**Pull request by Jeffwan**
## What changes were proposed in this pull request?
Changed the `kubernetes-client` version to 4.1.2.  Latest version fix error with exec credentials (used by aws eks) and this will be used to talk with kubernetes API server. Users can submit spark job to EKS api endpoint now with this patch. 

## How was this patch tested?
unit tests and manual tests.

### CODE REVIEW

**Comment by Jeffwan** [comment]
@dongjoon-hyun @ifilonenko @erikerlandson

**Comment by dongjoon-hyun** [comment]
Hi, @Jeffwan . Thank you for your first contribution.

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by Jeffwan** [comment]
I am trying to resolve integration test failure and submit revision later.

**Comment by skonto** [comment]
@Jeffwan I am already updating the version here: https://github.com/apache/spark/pull/23514

**Comment by Jeffwan** [comment]
Good to know. I didn't see PR links in original ticket and thought no one working on it.  @skonto. Could you help upgrade to v4.1.2?  There's several bug fixes important for spark to talking to kubernetes. We want to leverage this latest version. Once that's included, I can close this PR.

**Comment by skonto** [comment]
@Jeffwan I didnt see the ticket because my PR is a bit older but I needed that new version with the k8s updated model for long :) Anyway I can rebase mine if yours is merged. How do you want me to help? What doe snot work for you?

**Comment by rvesse** [comment]
@shaneknapp What version of K8S do the integration tests run against?  I notice from the compatibility matrix given by the Fabric8 folks that 4.1.1 and above requires K8S 1.9.0 or later.  If the integration tests are running against an older K8S version that could explain the test failures.

@Jeffwan @skonto A side effect of this change is that this will make the minimum supported K8S version 1.9.0.  Personally I think this is fine since 1.9 is already EOL and not receiving updates from the K8S community but we would need to make sure this is updated in the docs appropriately

**Comment by rvesse** [comment]
I would also note that a workaround for the AWS EKS issue is that user can first manually log in with AWS and then pass the OAuth token explicitly to `spark-submit` i.e.

screen shot 2019-02-19 at 11 02 26

(The above from a recent discussion on the #spark-operator K8S slack channel - https://kubernetes.slack.com/archives/CALBDHMTL/p1549469998419600?thread_ts=1549416963.416700&cid=CALBDHMTL)

**Comment by Jeffwan** [comment]
@rvesse Right, 1.9.0 is already sunset. Should be safe to use it now. The walkround will definitely work but token expired 20mins later. User doesn't like to refresh token frequently. This is not just for EKS, all the authentication using plugin to retrieve token will be impacted.  For long term, I think we need to have this patch

**Comment by Jeffwan** [comment]
@skonto It takes me sometime to reproduce integration test failure in my env, not sure what's broken there. Will update thread later.

**Comment by shaneknapp** [comment]
> @shaneknapp What version of K8S do the integration tests run against? I notice from the compatibility matrix given by the Fabric8 folks that 4.1.1 and above requires K8S 1.9.0 or later. If the integration tests are running against an older K8S version that could explain the test failures.
> 
right now we're testing against:

minikube v0.25.0
k8s 1.10.0

i have a "plan" in the not-so-distant-future to get us updated to:

minikube v0.34.1
k8s 1.13.2 (or newer)

> @Jeffwan @skonto A side effect of this change is that this will make the minimum supported K8S version 1.9.0. Personally I think this is fine since 1.9 is already EOL and not receiving updates from the K8S community but we would need to make sure this is updated in the docs appropriately

sgtm

**Comment by vanzin** [comment]
FYI, upgrading to 4.1.2 also fixes SPARK-25590, so that one could be closed when this change goes in.

**Comment by erikerlandson** [comment]
re: kube version support, our originally proposed policy was to support current version and previous, and so technically supporting 1.12 and 1.13 is now sufficient, although supporting more is obviously even better.
cc @liyinan926 @mccheah

**Comment by foxish** [comment]
Might make sense to match the k8s community and patch releases - 1.11 is still getting actively patched (for a total of 9 months since it came out). With the 3 month release cycle, we should maybe target 3 versions and deprecated 1.11 support only after 1.14 comes out (https://github.com/kubernetes/sig-release/tree/master/releases/release-1.14#tldr); which is the same schedule that k8s upstream follows.

**Comment by shaneknapp** [comment]
fyi i can't help out on this for the next ~2 weeks, as we're moving back in
to our lab after a major remodel.

it shouldn't be too much trouble to test against different k8s versions as
we can pass that in as an arg when starting minikube.

it will be a PITA to ensure that all of the pods, etc, are properly
configured to allow the tests to run.  for example, i'm currently testing
against minikube v0.33.4 and kubernetes 1.13 and running in to service
account/api access issues (which i will revisit once the dust settles here
in meatspace).

i would hesitate to have regularly scheduled integration tests for this
matrix, however, as we have limited physical resources available (number of
ubuntu build nodes).

if someone would like to open a jira for this, we could move the
conversation there.

**Comment by skonto** [comment]
I think supporting two out of the three mentioned here: https://kubernetes.io/docs/setup/version-skew-policy like 1.12 and 1.13 would be good enough. But that means running tests twice, I will create a jira @shaneknapp.

**Comment by foxish** [comment]
That version skew policy document talks about skew between various components in a k8s system. The release cycle maintains 3 releases at all times - for example, right now, 1.11, 1.12 and 1.13 are all actively maintained.

Here is the relevant doc:
https://github.com/kubernetes/sig-release/blob/master/releases/patch-releases.md

I am the current 1.11 release czar for k8s and will be cutting a release till 1.14 goes out. I don't mind going back to a policy with 2 releases supported if there's a good rationale for it but we would be discontinuing support for an actively parched and maintained version of K8s.

**Review by srowen — APPROVED** [review]
The version bump itself seems fine to me. Does it need to be backported?

**Comment by foxish** [comment]
Should be okay to backport into 2.4 I think. There's probably more EKS users than anyone running k8s 1.8 :) 

cc/ @skonto @erikerlandson

**Comment by skonto** [comment]
@foxish EKS is on [1.10 & 1.11](https://aws.amazon.com/eks/faqs) btw. So in general I guess the versioning is more critical when we do a release and trying to cover versions people actually use for good or bad at that point in time.

**Comment by skonto** [comment]
jira here: https://issues.apache.org/jira/browse/SPARK-26973

**Review by rvesse — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master. 2.4 is still on older major versions of the client, so I hesitate to back-port it directly. If it's essential, that probably needs another look and PR.

**Comment by vanzin** [comment]
Just wanted to point out that this was merged without a successful run of the integration tests. Now all integration tests are failing.

I see in the discussion that the new version is supposed to be compatible with k8s 1.9, and that seems to be the version running on jenkins (minikube 0.25 says it ships with k8s 1.9). Might be a bug in 4.1.2. Perhaps we should revert this until we figure out what's going on.

Anyway, I filed SPARK-26997; I might revert this patch if we don't figure out what's wrong soon.

**Comment by srowen** [comment]
Oh, I apologize, I now see those K8S tests are separate and failing. I'll revert it if it's not obvious how to fix-forward.

**Comment by shaneknapp** [comment]
1) we're running minikube 0.25.0 w/k8s 1.10.0

2) this bug is fucking weird and i just started looking in to it today.

3) pls to be reverting this merge for now thx

4) we're in the middle of moving back in to our lab after a massive remodel, so my available time is limited until the middle of next week

**Comment by vanzin** [comment]
Ok, I'll revert the patch.

**Comment by Jeffwan** [comment]
Sorry about that.. I have not figured it out and notice github doesn't show failures.. I will try to fix it soon.

**Comment by vanzin** [comment]
It does show failures: https://github.com/apache/spark/pull/23814#issuecomment-464534840

Maybe it should be a little bit louder, though, like the regular PR builder failures.

Anyway, people are looking at this as part of SPARK-26997, so if you have more info, please post it there.

**Comment by skonto** [comment]
@vanzin @srinathshankar there is a [compatibility matrix](https://github.com/fabric8io/kubernetes-client#compatibility-matrix) for the fabric8io client and the k8s support. 4.1.2 is not listed there so I am wondering if 1.10 is supported any more.

**Comment by shaneknapp** [comment]
> @vanzin @srowen there is a [compatibility matrix](https://github.com/fabric8io/kubernetes-client#compatibility-matrix) for the fabric8io client and the k8s support. 4.1.2 is not listed there so I am wondering if 1.10 is supported any more.

i believe this is the case.  i'm currently confirming this, and then looking in to upgrading minikube to something more recent.  our current version, v0.25.0 is only able to support up to k8s 1.10.0.

a minikube upgrade is not trivial.

**Comment by vanzin** [comment]
It's super weird to drop compatibility in .z releases, but hey, I'm not an RM on that project... :-/

**Comment by shaneknapp** [comment]
we're definitely incompatible w/4.1.2.

i'll need to do a pretty serious upgrade on our end, and hope that minikube is keeping up w/k8s mainline in a reasonable fashion.

more here:
https://issues.apache.org/jira/browse/SPARK-26997?focusedCommentId=16779599&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16779599

**Comment by shaneknapp** [comment]
TL;DR:  i got this to work!  we can deploy the infra changes next week, get this merged and move on.

https://issues.apache.org/jira/browse/SPARK-26997?focusedCommentId=16779881&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16779881

**Comment by Jeffwan** [comment]
@shaneknapp Thanks! Good to know it's a minikube issue. I was not sure integration failures and get stucked there.  Anyone working on these items now? If not, I will have a try to make it work

https://issues.apache.org/jira/browse/SPARK-26997?focusedCommentId=16780802&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16780802

**Comment by shaneknapp** [comment]
@Jeffwan we're tracking work here:  https://issues.apache.org/jira/browse/SPARK-26742

once we decide which version of k8s to test against, i can get the infra deployed simultaneously with the merging of both this PR and https://github.com/apache/spark/pull/23993.
